src/home_robot/home_robot/navigation_policy/language_navigation/languagenav_frontier_exploration_policy.py

```diff
@@ -0,0 +1,160 @@
+        return self.reach_single_category(
+            map_features, object_category, reject_visited_targets
+        )
+
+    def cluster_filtering(self, m):
+        # m is a 480x480 goal map
+        if not m.any():
+            return m
+        device = m.device
+
+        # cluster goal points
+        k = DBSCAN(eps=4, min_samples=1)
+        m = m.cpu().numpy()
+        data = np.array(m.nonzero()).T
+        k.fit(data)
+
+        # mask all points not in the largest cluster
+        mode = scipy.stats.mode(k.labels_, keepdims=True).mode.item()
+        mode_mask = (k.labels_ != mode).nonzero()
+        x = data[mode_mask]
+
+        m_filtered = np.copy(m)
+        m_filtered[x] = 0.0
+        m_filtered = torch.tensor(m_filtered, device=device)
+
+        return m_filtered
+
+    def reach_goal_if_in_map(
+        self,
+        map_features,
+        goal_category,
+        reject_visited_targets=False,
+    ):
+        """If the desired goal is in the semantic map, reach it."""
+        batch_size, _, height, width = map_features.shape
+        device = map_features.device
+
+        goal_map = torch.zeros((batch_size, height, width), device=device)
+        found_goal_current = torch.zeros(batch_size, dtype=torch.bool, device=device)
+
+        for e in range(batch_size):
+            # if the category goal was not found previously
+            if not found_goal_current[e]:
+                # the category to navigate to
+                category_map = map_features[
+                    e, goal_category[e] + 2 * MC.NON_SEM_CHANNELS, :, :
+                ]
+
+                if reject_visited_targets:
+                    # remove the target objects that the agent has already been close to
+                    category_map = category_map * (
+                        1 - map_features[e, MC.BLACKLISTED_TARGETS_MAP, :, :]
+                    )
+                # if the desired category is found with required constraints, set goal for navigation
+                if (category_map == 1).sum() > 0:
+                    goal_map[e] = category_map == 1
+                    found_goal_current[e] = True
+        return goal_map, found_goal_current
+
+    def get_frontier_map(self, map_features):
+        # Select unexplored area
+        if self.exploration_strategy == "seen_frontier":
+            frontier_map = (map_features[:, [MC.EXPLORED_MAP], :, :] == 0).float()
+        elif self.exploration_strategy == "been_close_to_frontier":
+            frontier_map = (map_features[:, [MC.BEEN_CLOSE_MAP], :, :] == 0).float()
+
+        # Dilate explored area
+        frontier_map = 1 - binary_dilation(
+            1 - frontier_map, self.dilate_explored_kernel
+        )
+
+        # Select the frontier
+        frontier_map = (
+            binary_dilation(frontier_map, self.select_border_kernel) - frontier_map
+        )
+        return frontier_map
+
+    def explore_otherwise(self, map_features, goal_map, found_goal):
```

Comment:
again i know we use this elsewhere - we should avoid duplication in order to make sure the code is easier to maintain

Reply:
maybe we need a "base_navigation_policy" class which can hold this kind of thing? @mukulkhanna